File: src/types/traditionArtifacts/sub/AnimalShapeSize.ts
import { Entity, Integer, Object, Required, String } from "tsondb/schema/dsl"
import { NestedTranslationMap } from "../../Locale.js"

export const AnimalShapeSize = Entity(import.meta.url, {
  name: "AnimalShapeSize",
  namePlural: "AnimalShapeSizes",
  type: () =>
    Object({
      volume: Required({
        comment: "The animal shape size’s volume points",
        type: Integer({ minimum: 1 }),
      }),
      ap_value: Required({
        comment: "The animal shape size’s adventure point value",
        type: Integer({ minimum: 1 }),
      }),
      translations: NestedTranslationMap(
        Required,
        "AnimalShapeSize",
        Object({
          name: Required({
            comment: "The animal shape size’s name.",
            type: String({ minLength: 1 }),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
})

File: Sources/OptolithDatabaseSchema/Types/TraditionArtifacts/Sub/AnimalShapeSize.swift
import FileDB

@Model
public struct AnimalShapeSize {
    /// The animal shape size’s volume points
    @Minimum(1)
    let volume: Int

    /// The animal shape size’s adventure point value
    @Minimum(1)
    let ap_value: Int

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [Locale.ID: Translation]

    @Embedded
    struct Translation {  // AnimalShapeSizeTranslation
        /// The animal shape size’s name.
        @MinLength(1)
        let name: String
    }
}
